core/ordersync_subprotocols.go

```diff
@@ -72,13 +72,19 @@ func (p *FilteredPaginationSubProtocol) HandleOrderSyncRequest(ctx context.Conte
 			return nil, fmt.Errorf("FilteredPaginationSubProtocol received request with wrong metadata type (got %T)", req.Metadata)
 		}
 	}
+
 	// It's possible that none of the orders in the current page match the filter.
 	// We don't want to respond with zero orders, so keep iterating until we find
 	// at least some orders that match the filter.
 	filteredOrders := []*zeroex.SignedOrder{}
 	var snapshotID string
-	var currentPage int
-	for currentPage = metadata.Page; len(filteredOrders) == 0; currentPage += 1 {
```

Comment:
The bug here is really subtle. `currentPage += 1` gets executed _before_ we check the exit condition `len(filteredOrders) == 0` for the loop . As a result, the returned `ordersync.Response` always had a `Page` that was one higher than it was supposed to be. The requester always requests the _next_ page (`response.Page + 1`) so this meant we were skipping every other page.

Reply:
Nice catch. This is pretty weird behavior for a for-loop, but it’s good to know.